{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/carlos/miniconda3/envs/gnn/lib/python3.11/site-packages/cooltools/lib/numutils.py:651: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit  # (nopython=True)\n",
      "/home/carlos/miniconda3/envs/gnn/lib/python3.11/site-packages/cooltools/lib/numutils.py:726: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit  # (nopython=True)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "np.seterr(divide=\"ignore\", invalid=\"ignore\")\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import os, subprocess\n",
    "\n",
    "import cooler\n",
    "import cooltools.lib.plotting\n",
    "from cooltools.api.saddle import saddle_strength\n",
    "\n",
    "import cooltools\n",
    "import bioframe\n",
    "import multiprocess as mp\n",
    "\n",
    "from tqdm import tqdm\n",
    "import multiprocessing\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "resolution = 100_000\n",
    "NPROC = 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "clr = cooler.Cooler(\n",
    "    f\"/home/carlos/Desktop/manuscripts/notebooks/matrices/t0_q30.mcool::/resolutions/{resolution}\"\n",
    ")\n",
    "\n",
    "view_df = pd.DataFrame(\n",
    "    {\n",
    "        \"chrom\": clr.chromnames,\n",
    "        \"start\": 0,\n",
    "        \"end\": clr.chromsizes.values,\n",
    "        \"name\": clr.chromnames,\n",
    "    }\n",
    ")\n",
    "bins = clr.bins()[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use RNAPII data for phasing\n",
    "\n",
    "import bbi\n",
    "\n",
    "RNAPII = bins[[\"chrom\", \"start\", \"end\"]].copy()\n",
    "stacks = bbi.stackup(\n",
    "    \"/home/carlos/Desktop/manuscripts/notebooks/matrices/ENCFF144IVU.bigWig\",\n",
    "    RNAPII[\"chrom\"],\n",
    "    RNAPII[\"start\"],\n",
    "    RNAPII[\"end\"],\n",
    "    bins=1,\n",
    ")\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "x_norm = MinMaxScaler().fit_transform(stacks)\n",
    "RNAPII.loc[:, \"RNAPII\"] = x_norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hg38_genome = bioframe.load_fasta('../matrices/GRCh38.fa')\n",
    "# gc_cov = bioframe.frac_gc(bins[['chrom', 'start', 'end']], hg38_genome)\n",
    "# gc_cov.to_csv('hg38_gc_cov_100kb.tsv',index=False,sep='\\t')\n",
    "\n",
    "GC = pd.read_csv(\"hg38_gc_cov_100kb.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RNASEQ = bins[[\"chrom\", \"start\", \"end\"]].copy()\n",
    "stacks_ = [bbi.stackup(\n",
    "    f\"/home/carlos/Desktop/manuscripts/notebooks/compartments/merged_t{sampleCode}.bw\",\n",
    "    RNASEQ[\"chrom\"],\n",
    "    RNASEQ[\"start\"],\n",
    "    RNASEQ[\"end\"],\n",
    "    bins=1,\n",
    ") for sampleCode in [\"00\", 12 , 30, 60]]\n",
    "\n",
    "for stack, sampleCode in zip(stacks_, [\"t0\", \"t12\", \"t30\", \"t60\"]):\n",
    "    x_norm = MinMaxScaler().fit_transform(stack)\n",
    "    RNASEQ.loc[:, sampleCode] = x_norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# phasing track tou use: RNAPII , RNASEQ or GC\n",
    "\n",
    "phasing_track = \"RNAPII\"\n",
    "\n",
    "if phasing_track == \"RNAPII\":\n",
    "    phasing = RNAPII.copy()\n",
    "elif phasing_track == \"GC\":\n",
    "    phasing = GC.copy()\n",
    "elif phasing_track == \"RNASEQ\":\n",
    "    phasing = RNASEQ.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "samples = [0, 12, 30, 60]\n",
    "eigen_data = []\n",
    "expected_ = []\n",
    "n_eig = 3\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "pbar = tqdm(total=len(samples))\n",
    "for sample_idx, sample in enumerate(samples):\n",
    "    clr = cooler.Cooler(\n",
    "        f\"/home/carlos/Desktop/manuscripts/notebooks/matrices/t{sample}_q30.mcool::/resolutions/{resolution}\"\n",
    "    )\n",
    "\n",
    "    map_ = mp.Pool(NPROC).map\n",
    "\n",
    "    if phasing_track == \"RNASEQ\":\n",
    "        phasing_now = phasing[[\"chrom\", \"start\", \"end\", f\"t{sample}\"]].copy()\n",
    "        cis_eigs = cooltools.eigs_cis(\n",
    "            clr, phasing_now, view_df=view_df, n_eigs=n_eig, map=map_, sort_metric=\"pearsonr\"\n",
    "        )\n",
    "    else:\n",
    "        cis_eigs = cooltools.eigs_cis(\n",
    "            clr, phasing, view_df=view_df, n_eigs=n_eig, map=map_, sort_metric=\"pearsonr\"\n",
    "        )\n",
    "    eigen_data.append(cis_eigs)\n",
    "\n",
    "    expected = cooltools.expected_cis(clr=clr, view_df=view_df, nproc=NPROC)\n",
    "    expected_.append(expected)\n",
    "    pbar.update(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e_vect_dfs = []\n",
    "for i, e_dfs_ in enumerate(eigen_data):    \n",
    "    for chrName in clr.chromnames:\n",
    "        e_vect = []\n",
    "        for eig_idx in range(1,4):\n",
    "            e_df = e_dfs_[1].loc[e_dfs_[1].chrom == chrName].copy()\n",
    "            e_vector = e_df[f\"E{eig_idx}\"]\n",
    "            non_na_bool = ~e_vector.isna().values\n",
    "\n",
    "            bw_now = phasing.loc[phasing[\"chrom\"] == chrName].copy()\n",
    "\n",
    "            e_vector = e_vector[non_na_bool]\n",
    "\n",
    "            if phasing_track == \"RNASEQ\":\n",
    "                bw_now = bw_now[f\"t{samples[i]}\"]\n",
    "                bw_now = bw_now[non_na_bool]\n",
    "            elif phasing_track == \"GC\":\n",
    "                bw_now = bw_now[\"GC\"][non_na_bool]\n",
    "            elif phasing_track == \"RNAPII\":\n",
    "                bw_now = bw_now[\"RNAPII\"][non_na_bool]\n",
    "\n",
    "            corr_Data = np.corrcoef(e_vector, bw_now)\n",
    "            corr = corr_Data[0,1]\n",
    "            e_vect.append((corr, chrName, f\"t{samples[i]}\", eig_idx))\n",
    "        \n",
    "        e_vect = sorted(e_vect, key=lambda x: x[0], reverse=True)\n",
    "\n",
    "        prev = []\n",
    "        for k in range(3):\n",
    "            for j in range(3):\n",
    "                if set((k,j)) in prev:\n",
    "                    continue\n",
    "                else:\n",
    "                    prev.append(set((k,j)))\n",
    "                if k != j:\n",
    "                    val = np.abs(e_vect[k][0] - e_vect[j][0])\n",
    "                    if val <= 0.1 and (k == 0 or j == 0):\n",
    "                        # print(sorted(e_vect, key=lambda x: x[0], reverse=True))\n",
    "                        # print(sorted(e_vect, key=lambda x: x[3], reverse=False))\n",
    "                        # print(k,j, val)\n",
    "                        if sorted(e_vect, key=lambda x: x[0], reverse=True) != sorted(e_vect, key=lambda x: x[3], reverse=False):\n",
    "                            print(\"!!!!!!!!!!!!! Not Equal !!!!!!!!!!!!!\")\n",
    "                            print(\"\\n\")\n",
    "\n",
    "        # make dataframe\n",
    "        e_vect = pd.DataFrame(e_vect, columns=[f\"{phasing_track}_corr\", \"chrom\", \"sample\", \"Eig_vector\"])\n",
    "        e_vect['Eig_vector'] = e_vect['Eig_vector'].astype(str) \n",
    "        e_vect['Eig_vector'] = [f\"E{eig}\" for eig in e_vect['Eig_vector']]\n",
    "        e_vect_dfs.append(e_vect)\n",
    "\n",
    "e_vect_dfs = pd.concat(e_vect_dfs)\n",
    "e_vect_dfs.to_csv(f\"evs/eigenvector_correlation_{phasing_track}.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_base = {'chrom': [chromName for chromName in clr.chromnames for _ in range(3)],\n",
    "              'Eig_vector': [f\"E{eig}\" for _ in range(len(clr.chromnames)) for eig in range(1,4)]\n",
    "}\n",
    "samples = [0, 12, 30, 60]\n",
    "df_base = pd.DataFrame(df_base)\n",
    "for sample in samples:\n",
    "    sampleName = f\"t{sample}\"\n",
    "    df_per_sample = df_base.copy()\n",
    "    for df_name in ['GC', 'RNASEQ', 'RNAPII']:\n",
    "        df = pd.read_csv(f\"evs/eigenvector_correlation_{df_name}.tsv\", sep=\"\\t\")\n",
    "        df = df[df['sample'] == sampleName]\n",
    "        df_per_sample[f\"{df_name}_corr\"] = df[f\"{df_name}_corr\"].values\n",
    "    \n",
    "    # reorder columns\n",
    "    df_per_sample = df_per_sample[['Eig_vector', 'chrom', 'GC_corr', 'RNASEQ_corr', 'RNAPII_corr']]\n",
    "    df_per_sample.to_csv(f\"evs/eigenvector_correlation_{sampleName}.tsv\", sep=\"\\t\", index=False)\n",
    "    df_per_sample.to_excel(f\"evs/eigenvector_correlation_{sampleName}.xlsx\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gnn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
